### ISSUE TITLE
Kamon Akka HTTP has a context propagation issue for Scala Futures

### ISSUE THREAD

**Issue by yiminyuan** [kamon-io/Kamon#919]
- Scala 2.12.12
- Kamon 2.1.9

We recently upgraded Kamon from 1.x to 2.1.9, and we observed that the current version of Kamon Akka HTTP has a context propagation issue for Scala Futures. We rely on Kamon Akka HTTP to propagate the HTTP headers and report Akka HTTP related metrics.

Let's say we have the pattern like this:
```scala
def runWithContextTag[T](customTags: TagSet)(invokeLaterFuture: => Future[T])(implicit ec: ExecutionContext): Future[T] = {
  val updatedContext = Kamon.currentContext().withTags(tags)
  // The logger in here can see the HTTP headers propagated by Kamon Akka HTTP
  val scope = Kamon.storeContext(updatedContext)
  // The logger in here can see the custom context tags in the newly stored context
  val future = invokeLaterFuture
  future.onComplete(_ => scope.close())
  future
}

def business: Future[Foo] = {
  // Scala chaining Futures, e.g., map, transform, recover, etc.
  // PROBLEM: The loggers in these Futures can NOT see the custom context tags
}

runWithContextTag(customTags)(business)
```
This context propagation issue is described in the code comment above. We didn't observe such issue in Kamon 1.x.

Even we disabled propagation and tracing, this context propagation issue still happens:
```
kamon.instrumentation.akka.http {
  server {
    propagation.enabled = no
    tracing.enabled = no
  }
  client {
    propagation.enabled = no
    tracing.enabled = no
  }
}
```
We also tried to leverage https://kamon.io/docs/latest/instrumentation/futures/ with `Kamon.runWithContextTag`, this context propagation issue happens too.

I guess some Future inceptions in Kamon Akka HTTP mess up context. It could be related to this known issue: https://github.com/kamon-io/Kamon/issues/846

**Temporary fix**
Do NOT import `kamon-akka-http`. Manually create a span and add HTTP headers and other custom context tags to the current context. The tradeoff is we won't be able to report Akka HTTP related metrics.

**Comment by alonsodomin** [comment]
We are facing the same issue, do we have any update on how to workaround it?

**Comment by danischroeter** [comment]
It looks like we face the same issue...

**Comment by ivantopo** [comment]
Hello @yiminyuan, quick update on this: 

We are about to release Kamon 2.2.0 which rolls back the Futures behavior to exactly how it was in Kamon 1.x, so this shouldn't be a problem anymore. There is a lot more info on #1021 and on the related PRs (#1033, #1034, and #1035).

**Comment by ivantopo** [comment]
Hey folks, probably some of the issues you saw here are now fixed with Kamon [2.2.0](https://github.com/kamon-io/Kamon/releases/tag/v2.2.0).

@yiminyuan since the main change in 2.2.0 is bringing Futures back to 1.x behavior, probably things are going to work out just fine for you when you upgrade. I wanted to add a few extra notes, though:
- In your original example, you are storing a `Context` in one thread and then closing the returned `Scope` in another thread (see that `future.onComplete(_ => scope.close())` you have there). That will lead to unexpected behavior before `Scope` instances are not thread safe. Scopes must be closed in the same thread where they were created.
- We recently helped @alonsodomin with some of they issues they had, and they were boiled down to I think 4 things things:
  - A similar issue with closing a scope in a different Thread from where it was created. 
  - Missing instrumentation for certain Cats-related Runnable implementations. See below.
  - A problem with initialization order between instrumentation modules. Fixed on 2.1.14.
  - Missing support for Tapir + Akka HTTP. Fixed on 2.1.17.

The cats issue was solved by adding this config:

```
kanela.modules {
  executor-service {
    within += "cats.effect.internals.Callback\\$AsyncIdempotentCallback"
  }
}
```

I think that should be it. Please try Kamon 2.2.0 and let us know if there are any new issues. Thanks for reporting!
